owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/SingleAPBandSteering.java

```diff
@@ -0,0 +1,272 @@
+	) {
+		short minRssi2G = DEFAULT_MIN_RSSI_2G;
+		short maxRssi2G = DEFAULT_MAX_RSSI_2G;
+		short minRssiNon2G = DEFAULT_MIN_RSSI_NON_2G;
+		long backoffTimeNs = DEFAULT_BACKOFF_TIME_NS;
+
+		String arg;
+		if ((arg = args.get("minRssi2G")) != null) {
+			minRssi2G = Short.parseShort(arg);
+		}
+		if ((arg = args.get("maxRssi2G")) != null) {
+			maxRssi2G = Short.parseShort(arg);
+		}
+		if ((arg = args.get("minRssiNon2G")) != null) {
+			minRssiNon2G = Short.parseShort(arg);
+		}
+		if ((arg = args.get("backoffTimeNs")) != null) {
+			backoffTimeNs = Short.parseShort(arg);
+		}
+
+		return new SingleAPBandSteering(
+			model,
+			zone,
+			deviceDataManager,
+			minRssi2G,
+			maxRssi2G,
+			minRssiNon2G,
+			backoffTimeNs
+		);
+	}
+
+	/** Constructor */
+	public SingleAPBandSteering(
+		DataModel model,
+		String zone,
+		DeviceDataManager deviceDataManager,
+		short minRssi2G,
+		short maxRssi2G,
+		short minRssiNon2G,
+		long backoffTimeNs
+	) {
+		super(model, zone, deviceDataManager);
+		this.minRssi2G = minRssi2G;
+		this.maxRssi2G = maxRssi2G;
+		this.minRssiNon2G = minRssiNon2G;
+		this.backoffTimeNs = backoffTimeNs;
+	}
+
+	@Override
+	public Map<String, Map<String, String>> computeApClientActionMap() {
+		Map<String, Map<String, String>> apClientActionMap = new HashMap<>();
+		// iterate through every AP
+		for (
+			Map.Entry<String, List<State>> entry : model.latestStates
+				.entrySet()
+		) {
+			// get the latest state
+			// TODO window size (look at multiple states)
+			// TODO window percent (% of samples that must violate thresholds)
+			List<State> states = entry.getValue();
+			if (states == null || states.isEmpty()) {
+				continue;
+			}
+			final String serialNumber = entry.getKey();
+			final State state = states.get(states.size() - 1);
+			// iterate through every radio and every connected client
+			if (state.interfaces == null || state.interfaces.length == 0) {
+				continue;
+			}
+			final long currentTimeNs = System.nanoTime();
+			for (State.Interface iface : state.interfaces) {
+				if (iface.ssids == null || iface.ssids.length == 0) {
+					continue;
+				}
+				for (State.Interface.SSID ssid : iface.ssids) {
+					if (
+						ssid.associations == null ||
+							ssid.associations.length == 0
+					) {
+						continue;
+					}
+					final State.Radio radio = gson.fromJson(
+						ssid.radio,
+						State.Radio.class
+					);
+					Map<String, Capabilities.Phy> capabilitiesPhy =
+						model.latestDeviceCapabilitiesPhy
+							.get(serialNumber);
+					if (capabilitiesPhy == null) {
+						continue;
+					}
+					final String band = ModelerUtils.getBand(
+						radio,
+						capabilitiesPhy
+					);
+					if (band == null) {
+						continue;
+					}
+					for (
+						State.Interface.SSID.Association assoc : ssid.associations
+					) {
+						// decide whether to do any band steering
+						// TODO check which bands AP & client can use (see 11k)
+						if (UCentralConstants.BAND_2G.equals(band)) {
```

Comment:
IMO this loop body is way too far nested and you should define a new method for this. (The code is line breaking too much to be readable)